### ISSUE TITLE
How to handle multiple applications with their own urlpatterns and different endpoint URLs?

### ISSUE THREAD

**Issue by debnet** [axnsan12/drf-yasg#661]
Everything in the title.

I have a Django project with several applications insides, each one "can be" standalone and have their own URL patterns.
I want to put a swagger endpoint for each application but how to achieve that?
In django-rest-swagger, I could specify an relative URL endpoint which was fine, but now OpenAPI generator seems to need an absolute one I can't provide for generic purposes.

Thanks.

**Comment by debnet** [comment]
Here the configuration I add to `my_app.urls` :

```python
    from drf_yasg2.views import get_schema_view
    from drf_yasg2 import openapi
    schema_view = get_schema_view(
        openapi.Info(
            title="Front APIs",
            default_version='v1',
        ),
        public=True,
        urlconf='front.api',
        # url='/api/front/',  # doesn't not work
    )
    urlpatterns += [
        url(r'^swagger(?P<format>\.json|\.yaml)$', schema_view.without_ui(cache_timeout=0), name='schema'),
        url(r'^swagger/$', schema_view.with_ui('swagger', cache_timeout=0), name='swagger'),
        url(r'^redoc/$', schema_view.with_ui('redoc', cache_timeout=0), name='redoc'),
    ]
```

Without this URL parameter, the APIs endpoints are incorrect because they don't include `/api/front/` prefix.

**Comment by axnsan12** [comment]
I'm not exactly sure I understand what is not working as expected here. The code as shown should be enough to generate a swagger page with only the `front.api` urls in it.

**Comment by debnet** [comment]
The issue here is in the URL: instead of having `http://localhost/api/front/<endpoint>` all the swagger calls are on the website root `http://localhost/<endpoint>/`, so listing of APIs with documentation works well, but testing doesn't work at all.

**Comment by axnsan12** [comment]
Is your /api/front prefix added by some proxy?

**Comment by debnet** [comment]
No, added by Django routing:

```python
    ...
    path("api/front/", include("front.api", namespace="front-api")),
    ...
```

**Comment by axnsan12** [comment]
Ok, then you will probably need to provide a complete, minimal project to reproduce this problem, otherwise I don't see how we can help.

**Comment by abenavidesmeli** [comment]
Yeah, I had that same problem,  it seem caused because the get_schema_view, it's placed in each app and not in the main app, that's why the root prefix seems to be omitted, I suppose that the behavior its correct because it's  like a api documentation per app.

What I did was create a custom generator  and override the get_schema in order to build the base_url properly and at this point use "Try it out" feature properly.

**Comment by Alkalit** [comment]
I really don't understand why such a simple and trivial thing should be so pain in ass. I simply want to set an absolute url path for a schema view. Without host and schema. That's all. 
I'm serving my app like this:
```
promise_map_swagger_view = get_schema_view(
    openapi.Info(
        title=SCHEMA_TITLE,
        default_version='v1'
    ),

    urlconf='apps.promise_map.urls',
    permission_classes=(permissions.AllowAny,),
)
```

```
        url(r'^api-pm/docs', promise_map_swagger_view.with_ui('swagger', cache_timeout=0)),
        url(r'^api/promise-map/', include('apps.promise_map.urls')),
```
And on the swagger page when I try those endpoints it shots them against `/`. I tried to move the swagger viewn in the promise-map.urls itself, but It didn't work.

**Comment by hubertokf** [comment]
same problem here

**Comment by mohitbestpeers** [comment]
> Yeah, I had that same problem, it seem caused because the get_schema_view, it's placed in each app and not in the main app, that's why the root prefix seems to be omitted, I suppose that the behavior its correct because it's like a api documentation per app, but at the same time it's generate a wrong base url, that's why all the request execute with the feature "Try it out" from the UI failed.
> 
> What I did was create a custom generator and override the get_schema in order to build the base_url properly and at this point use "Try it out" feature properly.

@abenavidesmeli , I want to prepend `/v1/` or some kind of prefix to it, so how can I able to achieve this? please help.

**Comment by MEmmanuel** [comment]
I personnally managed to get my API swagged with the base path, without explicitly specifying it in every path, by using a special `SchemaGenerator` like so (not really pretty, but it works) :

```python
swagger_base_path = "swagger"


class CustomOpenAPISchemaGenerator(OpenAPISchemaGenerator):
    def get_paths(self, endpoints, components, request, public):
        base_api_endpoint = request.path.split(swagger_base_path)[0].strip("/")
        fixed_endpoints = dict(
            (f"{base_api_endpoint}/{k}", v) for k, v in endpoints.items()
        )
        return super().get_paths(fixed_endpoints, components, request, public)
```

**Comment by mohitCodepy** [comment]
@MEmmanuel @mohitbestpeers , I've implemented that by overriding `OpenAPISchemaGenerator`, you can try like this


urls.py
```
class CustomSchemaGeneratorClass(OpenAPISchemaGenerator):
        def determine_path_prefix(self, request):
            return super().determine_path_prefix("/v1/")

schema_view = get_schema_view(
    openapi.Info(
        title="title",
        default_version="v1",
        description="desc",
    ),
    generator_class=CustomSchemaGeneratorClass,
)
```


custom_swagger.py
```
from drf_yasg.inspectors import SwaggerAutoSchema


class CustomSwaggerViewsetTag(SwaggerAutoSchema):

    """
    CustomSwaggerViewsetTag
        Adds the tag name to group similar routes in swagger.
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # get the url name after /v1/ and set as tag name
        self.tag_name = args[-1][1]

    def get_tags(self, operation_keys=None):
        tags = super().get_tags()
        tags[0] = self.tag_name
        return tags
```
and add to group your APIs according to app
swagger_schema = CustomSwaggerViewsetTag into your ModelViewSet
